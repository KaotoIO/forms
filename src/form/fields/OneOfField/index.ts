export * from './OneOfField';
export * from './SchemaList';
