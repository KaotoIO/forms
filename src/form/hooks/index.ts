export * from './field-value';
export * from './one-of-field';
