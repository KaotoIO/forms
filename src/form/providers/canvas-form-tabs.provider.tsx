import { FunctionComponent, PropsWithChildren, createContext, useState } from 'react';

type TabKeys = keyof typeof FormTabsModes;
export interface CanvasFormTabsContextResult {
  selectedTab: TabKeys;
  setSelectedTab: (key: TabKeys) => void;
}

export const CanvasFormTabsContext = createContext<CanvasFormTabsContextResult>({
  selectedTab: 'Required',
  setSelectedTab: () => void 0,
});

/**
 * Used for fetching and injecting the selected tab information from the canvas form
 */
export const CanvasFormTabsProvider: FunctionComponent<PropsWithChildren<{ tab?: TabKeys }>> = ({
  tab = 'Required',
  children,
}) => {
  const [selectedTab, setSelectedTab] = useState<TabKeys>(tab);

  return (
    <CanvasFormTabsContext.Provider
      value={{
        selectedTab,
        setSelectedTab,
      }}
    >
      {children}
    </CanvasFormTabsContext.Provider>
  );
};

export const FormTabsModes = {
  Required: 'Shows Required fields only',
  All: 'Shows All fields',
  Modified: 'Shows Modified fields only',
};
