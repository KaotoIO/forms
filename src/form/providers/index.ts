export * from './canvas-form-tabs.provider';
export * from './filtered-field.provider';
export * from './FormComponentFactoryProvider';
export * from './ModelProvider';
export * from './SchemaDefinitionsProvider';
export * from './SchemaProvider';
export * from './SuggestionRegistryProvider';
