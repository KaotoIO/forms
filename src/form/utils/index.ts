export * from './camel-case-to-space';
export * from './camel-random-id';
export * from './get-field-groups';
export * from './get-filtered-properties';
export * from './get-item-from-schema';
export * from './get-value';
export * from './is-defined';
export * from './is-raw-string';
export * from './resolve-schema-with-ref';
export * from './set-value';
