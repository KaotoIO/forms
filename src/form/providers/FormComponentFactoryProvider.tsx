import { JSONSchema4 } from 'json-schema';
import { FunctionComponent, PropsWithChildren, useCallback } from 'react';
import { ArrayField } from '../fields/ArrayField/ArrayField';
import { BooleanField } from '../fields/BooleanField';
import { DisabledField } from '../fields/DisabledField';
import { EnumField } from '../fields/EnumField';
import { AllOfField } from '../fields/ObjectField/AllOfField';
import { ObjectField } from '../fields/ObjectField/ObjectField';
import { OneOfField } from '../fields/OneOfField/OneOfField';
import { PasswordField } from '../fields/PasswordField';
import { PropertiesField } from '../fields/PropertiesField/PropertiesField';
import { StringField } from '../fields/StringField';
import { TextAreaField } from '../fields/TextAreaField';
import { FieldProps } from '../models/typings';
import {
  FormComponentFactoryContext,
  FormComponentFactoryContextValue,
} from './context/form-component-factory-context';

export type CustomFieldsFactory = (schema: JSONSchema4) => FunctionComponent<FieldProps> | undefined;
interface IFormComponentFactoryProvider extends PropsWithChildren {
  /**
   * Factory function to override the Field selection algorithm.
   *
   * Should return a component receiving FieldProps as props, or undefined if no custom field is found.
   */
  customFieldsFactory?: CustomFieldsFactory;
}
export const FormComponentFactoryProvider: FunctionComponent<IFormComponentFactoryProvider> = ({
  customFieldsFactory,
  children,
}) => {
  const factory = useCallback<FormComponentFactoryContextValue>(
    (schema) => {
      const CustomField = customFieldsFactory?.(schema);
      if (CustomField) {
        return CustomField;
      }

      if (schema.format === 'password') {
        return PasswordField;
      } else if (schema.type === 'string' && schema.title && TEXT_AREA_PROPERTY_NAMES.includes(schema.title)) {
        return TextAreaField;
      } else if (schema.type === 'string' && Array.isArray(schema.enum)) {
        return EnumField;
      } else if (schema.type === 'object' && Object.keys(schema?.properties ?? {}).length === 0) {
        /*
         * If the object has no properties, we render a generic key-value pairs field
         * This is useful for langchain4j-tools consumer components or when configuring beans entities
         */
        return PropertiesField;
      }

      switch (schema.type) {
        case 'string':
        case 'number':
        case 'integer':
          return StringField;
        case 'boolean':
          return BooleanField;
        case 'object':
          return ObjectField;
        case 'array':
          return ArrayField;
      }

      if (Array.isArray(schema.oneOf)) {
        return OneOfField;
      } else if (Array.isArray(schema.allOf)) {
        return AllOfField;
      } else if (Array.isArray(schema.anyOf)) {
        throw new Error('FormComponentFactoryProvider: AnyOf should be handled in the scope of the ObjectField');
      }

      return DisabledField;
    },
    [customFieldsFactory],
  );

  return <FormComponentFactoryContext.Provider value={factory}>{children}</FormComponentFactoryContext.Provider>;
};

/* Name of the properties that should load TextAreaField */
const TEXT_AREA_PROPERTY_NAMES = ['Expression', 'Description', 'Query', 'Script'];
