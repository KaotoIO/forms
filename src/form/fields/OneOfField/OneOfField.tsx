import { FunctionComponent, useContext } from 'react';
import { useOneOfField } from '../../hooks/one-of-field';
import { SchemaContext, SchemaProvider } from '../../providers/SchemaProvider';
import { FieldProps } from '../../models/typings';
import { AutoField } from '../AutoField';
import { SchemaList } from './SchemaList';
import { ArrayFieldWrapper } from '../ArrayField/ArrayFieldWrapper';

export const OneOfField: FunctionComponent<FieldProps> = ({ propName, required }) => {
  const { selectedOneOfSchema, oneOfSchemas, onSchemaChange, shouldRender } = useOneOfField(propName);
  const { schema } = useContext(SchemaContext);
  let title = 'Type';
  switch (schema.format) {
    case 'dataFormatType':
      title = 'Data Format Type';
      break;
    case 'loadBalancerType':
      title = 'Load Balancer Type';
      break;
    case 'expression':
    case 'expressionProperty':
      title = 'Expression';
      break;
    case 'errorHandlerType':
      title = 'Error Handler Type';
      break;
  }

  const onCleanInput = () => {
    onSchemaChange();
  };

  if (!shouldRender) {
    return null;
  }

  return (
    <ArrayFieldWrapper
      propName={propName}
      type="expression"
      title={selectedOneOfSchema?.name ?? title}
      description={selectedOneOfSchema?.description}
      actions={
        <SchemaList
          aria-label={`${propName} oneof list`}
          data-testid={`${propName}__oneof-list`}
          propName={propName}
          selectedSchema={selectedOneOfSchema}
          schemas={oneOfSchemas}
          onChange={onSchemaChange}
          onCleanInput={onCleanInput}
          placeholder={`Select or write the ${title}`}
        />
      }
    >
      {selectedOneOfSchema && (
        <SchemaProvider schema={selectedOneOfSchema.schema}>
          <AutoField propName={propName} required={required} />
        </SchemaProvider>
      )}
    </ArrayFieldWrapper>
  );
};
