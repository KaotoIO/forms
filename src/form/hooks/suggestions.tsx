import { Menu, MenuContent, MenuItem, MenuList, Popper } from '@patternfly/react-core';
import { JSONSchema4 } from 'json-schema';
import { ReactNode, RefObject, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { GroupedSuggestions, Suggestion, SuggestionProvider } from '../models/suggestions';
import { SuggestionContext } from '../providers';
import { applySuggestion } from '../utils/apply-suggestion';
import { getCursorWord } from '../utils/get-cursor-word';

type UseSuggestionsProps = {
  propName: string;
  schema: JSONSchema4;
  inputRef: RefObject<HTMLInputElement>;
  value: string | number;
  setValue?: (value: string | number) => void;
};
export const useSuggestions = ({ propName, schema, inputRef, value, setValue }: UseSuggestionsProps): ReactNode => {
  const [isVisible, setIsVisible] = useState(false);
  const [groupedSuggestions, setGroupedSuggestions] = useState<GroupedSuggestions>({ root: [] });
  const menuRef = useRef<HTMLDivElement>(null);
  const firstElementRef = useRef<HTMLDivElement>(null);

  const { getProviders } = useContext(SuggestionContext);

  const suggestionProviders: SuggestionProvider[] = useMemo(
    () => getProviders(propName, schema),
    [getProviders, propName, schema],
  );

  const onEscapeKey = useCallback(
    (event: React.KeyboardEvent | KeyboardEvent) => {
      event.preventDefault();
      setIsVisible(false);
      requestAnimationFrame(() => {
        inputRef.current?.focus();
      });
    },
    [inputRef],
  );

  const handleInputKeyDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.ctrlKey && event.code === 'Space') {
        event.preventDefault();
        setIsVisible(true);
        requestAnimationFrame(() => {
          firstElementRef.current?.focus();
          firstElementRef.current?.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'nearest' });
        });
      } else if (event.key === 'Escape') {
        onEscapeKey(event);
      }
    },
    [onEscapeKey],
  );

  const getHandleOnClick = useCallback(
    (inputValue: string | number, suggestion: Suggestion) => () => {
      const { newValue, cursorPosition } = applySuggestion(suggestion, inputValue, inputRef.current?.selectionStart);

      setIsVisible(false);
      setValue?.(newValue);

      inputRef.current?.focus();
      inputRef.current?.setSelectionRange(cursorPosition, cursorPosition);
    },
    [inputRef, setValue],
  );

  const getHandleMenuKeyDown = useCallback(
    (inputValue: string | number, suggestion?: Suggestion) => (event: React.KeyboardEvent) => {
      if (event.key === 'Enter' && suggestion) {
        event.preventDefault();
        getHandleOnClick(inputValue, suggestion)();
      } else if (event.key === 'Escape') {
        onEscapeKey(event);
      }
    },
    [getHandleOnClick, onEscapeKey],
  );

  /** Fetch suggestions from providers */
  useEffect(() => {
    let cancelled = false;
    if (!isVisible) return;

    const fetchSuggestions = async () => {
      const cursorPosition = inputRef.current?.selectionStart;
      const { word } = getCursorWord(value, cursorPosition);

      const results = await Promise.all(
        suggestionProviders.map((provider) =>
          Promise.resolve(
            provider.getSuggestions(word, {
              propertyName: propName,
              inputValue: value,
              cursorPosition,
            }),
          ),
        ),
      );

      if (cancelled) return;

      const newGroupedSuggestions = results.flat().reduce(
        (acc, suggestion) => {
          const group = suggestion.group ?? 'root';
          acc[group] ??= [];
          acc[group].push(suggestion);
          return acc;
        },
        { root: [] } as GroupedSuggestions,
      );
      setGroupedSuggestions(newGroupedSuggestions);
    };

    fetchSuggestions();
    return () => {
      cancelled = true;
    };
  }, [suggestionProviders, value, propName, inputRef, isVisible]);

  /** Register keyboard bindings */
  useEffect(() => {
    const input = inputRef.current;
    if (!input) return;

    const handleFocus = () => {
      input.addEventListener('keydown', handleInputKeyDown);
    };
    const handleBlur = () => {
      input.removeEventListener('keydown', handleInputKeyDown);
    };

    input.addEventListener('focus', handleFocus);
    input.addEventListener('blur', handleBlur);

    // If already focused, register immediately
    if (document.activeElement === input) {
      handleFocus();
    }

    return () => {
      input.removeEventListener('focus', handleFocus);
      input.removeEventListener('blur', handleBlur);
      input.removeEventListener('keydown', handleInputKeyDown);
    };
  }, [handleInputKeyDown, inputRef]);

  return (
    <Popper
      isVisible={isVisible}
      triggerRef={inputRef}
      popperRef={menuRef}
      popper={
        <div ref={menuRef}>
          <Menu containsFlyout data-testid="suggestions-menu">
            <MenuContent>
              <MenuList>
                {Object.entries(groupedSuggestions).map(([group, suggestions], groupIndex) => {
                  if (group === 'root') {
                    return suggestions.map((suggestion, suggestionIndex) => {
                      const isFirst = groupIndex === 0 && suggestionIndex === 0;
                      return (
                        <MenuItem
                          ref={isFirst ? firstElementRef : null}
                          key={suggestion.value}
                          onClick={getHandleOnClick(value, suggestion)}
                          onKeyDown={getHandleMenuKeyDown(value, suggestion)}
                          description={suggestion.description}
                        >
                          {suggestion.value}
                        </MenuItem>
                      );
                    });
                  }

                  return (
                    <MenuItem
                      ref={groupIndex === 0 ? firstElementRef : null}
                      key={group}
                      flyoutMenu={
                        <Menu containsFlyout>
                          <MenuContent>
                            <MenuList>
                              {suggestions.map((suggestion) => (
                                <MenuItem
                                  key={suggestion.value}
                                  onClick={getHandleOnClick(value, suggestion)}
                                  onKeyDown={getHandleMenuKeyDown(value, suggestion)}
                                  description={suggestion.description}
                                >
                                  {suggestion.value}
                                </MenuItem>
                              ))}
                            </MenuList>
                          </MenuContent>
                        </Menu>
                      }
                    >
                      <strong>{group}</strong>
                    </MenuItem>
                  );
                })}

                {groupedSuggestions.root.length === 0 && <MenuItem isDisabled>No suggestions available</MenuItem>}
              </MenuList>
            </MenuContent>
          </Menu>
        </div>
      }
    />
  );
};
