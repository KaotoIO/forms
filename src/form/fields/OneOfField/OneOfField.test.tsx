import { render, screen, fireEvent } from '@testing-library/react';
import { CanvasFormTabsContextResult } from '../../providers';
import { FormComponentFactoryProvider } from '../../providers/FormComponentFactoryProvider';
import { SchemaProvider } from '../../providers/SchemaProvider';
import { OneOfSchemas, ROOT_PATH } from '../../utils';
import { OneOfField } from './OneOfField';

// Mock the useOneOfField hook
jest.mock('../../hooks/one-of-field', () => ({
  useOneOfField: jest.fn(),
}));

import { useOneOfField } from '../../hooks/one-of-field';

const mockUseOneOfField = useOneOfField as jest.MockedFunction<typeof useOneOfField>;

describe('OneOfField', () => {
  beforeEach(() => {
    // Reset mock
    jest.clearAllMocks();
  });

  const mockOneOfSchemas = [
    {
      name: 'Test Schema 1',
      description: 'First test schema',
      schema: { type: 'object', properties: { name: { type: 'string' } } },
    },
    {
      name: 'Test Schema 2',
      description: 'Second test schema',
      schema: { type: 'object', properties: { value: { type: 'number' } } },
    },
  ] as OneOfSchemas[];

  const defaultMockReturn = {
    selectedOneOfSchema: mockOneOfSchemas[0],
    oneOfSchemas: mockOneOfSchemas,
    onSchemaChange: jest.fn(),
    shouldRender: true,
  };

  const schema = {
    oneOf: mockOneOfSchemas,
  };

  it('should render correctly with basic oneOf schema', () => {
    mockUseOneOfField.mockReturnValue(defaultMockReturn);

    const { container } = render(
      <FormComponentFactoryProvider>
        <SchemaProvider schema={schema}>
          <OneOfField propName={ROOT_PATH} />
        </SchemaProvider>
      </FormComponentFactoryProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render with different format types and set appropriate titles', () => {
    const testCases = [
      { format: 'dataFormatType', expectedTitle: 'Data Format Type' },
      { format: 'loadBalancerType', expectedTitle: 'Load Balancer Type' },
      { format: 'expression', expectedTitle: 'Expression' },
      { format: 'errorHandlerType', expectedTitle: 'Error Handler Type' },
    ];

    testCases.forEach(({ format, expectedTitle }) => {
      // Mock with no selectedOneOfSchema to test the format-based title logic
      mockUseOneOfField.mockReturnValue({
        ...defaultMockReturn,
        selectedOneOfSchema: undefined,
      });

      schema['format'] = format;

      const wrapper = render(
        <FormComponentFactoryProvider>
          <SchemaProvider schema={schema}>
            <OneOfField propName={ROOT_PATH} />
          </SchemaProvider>
        </FormComponentFactoryProvider>,
      );

      expect(wrapper?.getByText(expectedTitle)).toBeInTheDocument();
    });
  });

  it('should not render when shouldRender is false', () => {
    mockUseOneOfField.mockReturnValue({
      ...defaultMockReturn,
      shouldRender: false,
    });

    const { container } = render(
      <FormComponentFactoryProvider>
        <SchemaProvider schema={schema}>
          <OneOfField propName={ROOT_PATH} />
        </SchemaProvider>
      </FormComponentFactoryProvider>,
    );

    expect(container.firstChild).toBeNull();
  });

  it('should handle schema selection and change functionality', () => {
    const mockOnSchemaChange = jest.fn();
    mockUseOneOfField.mockReturnValue({
      ...defaultMockReturn,
      onSchemaChange: mockOnSchemaChange,
    });

    render(
      <FormComponentFactoryProvider>
        <SchemaProvider schema={schema}>
          <OneOfField propName={ROOT_PATH} />
        </SchemaProvider>
      </FormComponentFactoryProvider>,
    );

    // Test schema change
    fireEvent.click(screen.getByRole('button', { name: 'Test Schema 2' }));
    expect(mockOnSchemaChange).toHaveBeenCalledWith(mockOneOfSchemas[1]);
  });
});
