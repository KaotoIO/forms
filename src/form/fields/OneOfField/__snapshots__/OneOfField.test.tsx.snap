// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OneOfField should render correctly with basic oneOf schema 1`] = `
<div>
  <div
    class="pf-v6-c-card"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF6/Card"
    data-ouia-safe="true"
    data-testid="#__field-wrapper"
    id=""
  >
    <div
      class="pf-v6-c-card__header"
    >
      <div
        class="pf-v6-c-card__actions"
      >
        <div
          aria-labelledby="#-legend"
          class="pf-v6-c-form__group"
          role="group"
        >
          <div
            class="pf-v6-c-form__group-control pf-m-stack"
          >
            <div
              aria-label="# oneof list"
              class="pf-v6-c-toggle-group"
              data-testid="#__oneof-list"
              id="#"
              role="group"
            >
              <div
                class="pf-v6-c-toggle-group__item"
              >
                <button
                  aria-pressed="true"
                  class="pf-v6-c-toggle-group__button pf-m-selected"
                  id="Test Schema 1"
                  type="button"
                >
                  <span
                    class="pf-v6-c-toggle-group__text"
                  >
                    Test Schema 1
                  </span>
                </button>
              </div>
              <div
                class="pf-v6-c-toggle-group__item"
              >
                <button
                  aria-pressed="false"
                  class="pf-v6-c-toggle-group__button"
                  id="Test Schema 2"
                  type="button"
                >
                  <span
                    class="pf-v6-c-toggle-group__text"
                  >
                    Test Schema 2
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="pf-v6-c-card__header-main"
      >
        <div
          class="pf-v6-c-card__title"
        >
          <div
            class="pf-v6-c-card__title-text"
          >
             
            Test Schema 1
             
            <div
              style="display: contents;"
            >
              <span
                aria-label="More info for Test Schema 1 field"
                class="pf-v6-c-button pf-m-plain pf-m-no-padding"
                data-ouia-component-id="OUIA-Generated-Button-plain-1"
                data-ouia-component-type="PF6/Button"
                data-ouia-safe="true"
                role="button"
                tabindex="0"
                type="button"
              >
                <span
                  class="pf-v6-c-button__icon"
                >
                  <svg
                    aria-hidden="true"
                    class="pf-v6-svg"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M521.3,576 C627.5,576 713.7,502 713.7,413.7 C713.7,325.4 627.6,253.6 521.3,253.6 C366,253.6 334.5,337.7 329.2,407.2 C329.2,414.3 335.2,416 343.5,416 L445,416 C450.5,416 458,415.5 460.8,406.5 C460.8,362.6 582.9,357.1 582.9,413.6 C582.9,441.9 556.2,470.9 521.3,473 C486.4,475.1 447.3,479.8 447.3,521.7 L447.3,553.8 C447.3,570.8 456.1,576 472,576 C487.9,576 521.3,576 521.3,576 M575.3,751.3 L575.3,655.3 C575.313862,651.055109 573.620137,646.982962 570.6,644 C567.638831,640.947672 563.552355,639.247987 559.3,639.29884 L463.3,639.29884 C459.055109,639.286138 454.982962,640.979863 452,644 C448.947672,646.961169 447.247987,651.047645 447.29884,655.3 L447.29884,751.3 C447.286138,755.544891 448.979863,759.617038 452,762.6 C454.961169,765.652328 459.047645,767.352013 463.3,767.30116 L559.3,767.30116 C563.544891,767.313862 567.617038,765.620137 570.6,762.6 C573.659349,759.643612 575.360354,755.553963 575.3,751.3 M512,896 C300.2,896 128,723.9 128,512 C128,300.3 300.2,128 512,128 C723.8,128 896,300.2 896,512 C896,723.8 723.7,896 512,896 M512.1,0 C229.7,0 0,229.8 0,512 C0,794.2 229.8,1024 512.1,1024 C794.4,1024 1024,794.3 1024,512 C1024,229.7 794.4,0 512.1,0"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v6-c-card__body pf-v6-c-form kaoto-form__label"
      data-testid="#__children"
    />
  </div>
</div>
`;
